sdk/storage/azure-storage-common/src/storage_bearer_token_authentication_policy.cpp

```diff
@@ -0,0 +1,51 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#include "azure/storage/common/internal/storage_bearer_token_authentication_policy.hpp"
+
+#include "azure/storage/common/internal/constants.hpp"
+
+#include <azure/core/internal/credentials/authorization_challenge_parser.hpp>
+
+namespace Azure { namespace Storage { namespace _internal {
+
+  std::unique_ptr<Azure::Core::Http::RawResponse>
+  StorageBearerTokenAuthenticationPolicy::AuthorizeAndSendRequest(
+      Azure::Core::Http::Request& request,
+      Azure::Core::Http::Policies::NextHttpPolicy& nextPolicy,
+      Azure::Core::Context const& context) const
+  {
+    std::string tenantId = m_safeTenantId.Get();
+    if (!tenantId.empty() || !m_enableTenantDiscovery)
```

Comment:
Does this condition work as you intended? Did you mean `&&`?
* If tenant discovery is disabled, it is still going to override tenantId with an empty one. In which case it is going to get the challenge response, and your `AuthorizeRequestOnChallenge()` method will be called.
* Unless authentication was successful, `AuthorizeRequestOnChallenge()` will be called when we get a challenge response, even if you did not override tenantId in `AuthorizeAndSendRequest()`, so it is going to set `m_tenantId`, and unless it returns `false`, another attempt will be made to get a token, with the new tenantId.
* Next time it comes to authorization, your `m_tenantId` is not empty, so you ignore `m_enableTenantDiscovery`.

Reply:
Yeah, this condition is generally correct. 

And the reason why I add `EnableTenantDiscovery` is that the old logic will attmpt to get a token from AAD before real request, so if tenantId is not set in `ClientSecretCredential`, a `404` error from AAD will be received, and can't trigger any bearer challenge. 

And when `EnableTenantDiscovery` is true, it will directly send request without token, and received a 401 bearer challange. With the tenantId from challange, attempt will be made to get a token. So even if the tenantId is empty in `ClientSecretCredential`, it can get the correct tenantId from bearer challenge.